### PR TITLE
raftstore: make apply before persist compatible with online unsafe recovery

### LINKED ISSUE #16796: Apply raft log before persistence is not compatible with online unsafe recovery
## Bug Report

When feature "apply raft log before persistence" is enable, it is possible that applied_index > last_index, in this scenario, online unsafe recovery make not handle the recovery correctly.

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: ref #16717, close #16796

The unit test in this pr depends on #16626

What's Changed:

```commit-message
Before enter force-leader, if local applied_index > last_index, first force commit and truncate raft index to applied index. This can ensure the raft state is compatible with apply_state.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/peer.rs [hunk 1]
```diff
@@ -1647,6 +1652,26 @@ where
             return;
         }
 
+        // apply is ahead of raft, need to force raft commit to applied index.
```
Comment (v01dstar) [inline components/raftstore/src/store/fsm/peer.rs:1655]: Can you please add a comment that explains under what circumstances, applied index could be larger than last commit?

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -4365,6 +4386,30 @@ where
         }
     }
 
+    fn force_compact(&mut self, ctx: &mut ApplyContext<EK>, term: u64, compact_index: u64) {
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/apply.rs:4389]: maybe call it `unsafe_force_compact`

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -4365,6 +4386,30 @@ where
...
+        assert_eq!(self.delegate.apply_state.applied_index, compact_index);
+        if self.delegate.apply_state.get_truncated_state().index != compact_index {
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/apply.rs:4391]: `!=` -> `<`?

File: components/raftstore/src/store/fsm/peer.rs [hunk 4]
```diff
@@ -5247,6 +5273,53 @@ where
...
+                        peer_store.clear_entry_cache_warmup_state();
+                        peer_store.compact_entry_cache(last_index + 1);
+                        peer_store.raft_state_mut().mut_hard_state().commit = last_index;
+                        peer_store.raft_state_mut().last_index = last_index;
+                    }
+
+                    let raft_engine = self.fsm.peer.get_store().raft_engine();
+                    let mut batch = raft_engine.log_batch(2);
+                    raft_engine
+                        .gc(self.region_id(), first_index, last_index, &mut batch)
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/peer.rs:5294]: shouldn't introduce other io operation in raftstore thread

**Review by Connor1996 — APPROVED** [review]
LGTM

File: tests/failpoints/cases/test_unsafe_recovery.rs [hunk 5]
```diff
@@ -547,3 +552,169 @@ fn test_unsafe_recovery_rollback_merge() {
 
     fail::remove("on_schedule_merge_ret_err");
 }
+
+// Test the compatibility between apply unperisisted raft log with unsafe
```
Comment (Connor1996) [inline tests/failpoints/cases/test_unsafe_recovery.rs:556]:
```suggestion
// test the compatibility between apply before persist with unsafe recovery.
```

File: components/raftstore/src/store/fsm/peer.rs (hunk 4 above)
Comment (glorv) [reply components/raftstore/src/store/fsm/peer.rs:5289]: After discussing offline, we decide to keep the current code to make the logic simpler. We maybe enhance it in the future if this operation is still a thing.

**Review by v01dstar — APPROVED** [review]
LGTM

**Pull request merged**